File: Src/Runtime/Lists/Join.cpp
/*
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #                           Typed                     # *
 * #                   A text formatting DSL             # *
 * #                                                     # *
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #         Created by Rodrigo R. & Contributors        # *
 * #         Released under the Apache License 2.0       # *
 * #            Check LICENSE.MD for more info           # *
 * #                                                     # *
 * #-----------------------------------------------------# *
*/

//
// Created by Rodrigo on 7/16/26.
//

#include "ADT/Lang/AST.h"
#include "ADT/List/Object.h"
#include "Join.h"

#include "ADT/Exception/MismatchedType.h"
#include "Support/Format/Boolean.h"
#include "Support/Format/Dictionary.h"
#include "Support/Format/Float.h"
#include "Support/Format/Integer.h"
#include "Support/Format/List.h"
#include "Support/Format/String.h"
#include "Support/Printer/ASTPrinter.h"
#include "Support/Runtime/AccessString.h"

using namespace Typed;
using namespace Typed::Runtime;
using namespace Typed::Runtime::Lists;

ADT::Runtime::Object Lists::Join(
    ADT::List::Object &args,
    ADT::Lang::AST *trace
)
{
    auto &list = Support::Runtime::GetListObj(args[0]);
    auto delim = Support::Runtime::AccessString(args[1]);
    Celery::Str::String result;

    // Transverse the list
    for (auto &item : *list)
    {
        switch (item.type)
        {
            case ADT::Runtime::ObjectType::String:
            {
                Support::Format::String(item, result);
                break;
            }

            case ADT::Runtime::ObjectType::Integer:
            {
                Support::Format::Integer(item, result);
                break;
            }

            case ADT::Runtime::ObjectType::Float:
            {
                Support::Format::Float(item, result);
                break;
            }

            case ADT::Runtime::ObjectType::Boolean:
            {
                Support::Format::Boolean(item, result);
                break;
            }

            case ADT::Runtime::ObjectType::List:
            {
                Support::Format::List(item, result);
                break;
            }

            case ADT::Runtime::ObjectType::Dictionary:
            {
                Support::Format::Dictionary(item, result);
                break;
            }

            default:
            {
                result.Write("[Non_Formattable]", 17);
                break;
            }
        }
    }

    return {
        ADT::Runtime::ObjectType::String,
        std::move(result)
    };
}

File: src/runtime/lists/join.rs
/*
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #                           Typed                     # *
 * #                   A text formatting DSL             # *
 * #                                                     # *
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #         Created by Rodrigo R. & Contributors        # *
 * #         Released under the Apache License 2.0       # *
 * #            Check LICENSE.MD for more info           # *
 * #                                                     # *
 * #-----------------------------------------------------# *
*/
use std::cmp::max;
use std::fmt::Write;
use crate::adt::error::RuntimeError;
use crate::adt::lang::{ASTValue, Kind, RuntimeArguments, AST};
use crate::adt::result::ExecutionTupleResult;
use crate::adt::runtime::{HashableObject, Object};
use crate::runtime::strings::format::format_obj;
use crate::support::runtime::execution::continue_execution;
use crate::support::runtime::object::{get_list, get_string};

pub fn join<'a>(
    args: RuntimeArguments<'a>,
    trace: &AST<'a>
)
    -> ExecutionTupleResult<'a>
{
    let list = args.get(0).unwrap();
    let delim = args.get(1).unwrap();

    let delim = get_string(delim, trace)?;
    let list = get_list(list, trace)?;
    let list = list.borrow();
    let size = max(list.len(), 1) - 1;
    
    let mut result = String::new();
    
    for (i, item) in list.iter().enumerate() {
        format_obj(
            item,
            Kind::Any,
            trace,
            &mut result
        )?;
        
        if i != size {
            let res = result.write_str(delim);
            
            if res.is_err() {
                return Err(
                    RuntimeError::could_not_write(trace)
                )
            }
        }
    }

    continue_execution(
        Object::Hashable(
            HashableObject::String(
                ASTValue::Owned(result)
            )
        )
    )
}